### PR TITLE
Devops/1032/scheduled owasp

### LINKED ISSUE #1032: Perform scheduled OWASP scans against deployed site(s)
**Description:**
As outlined in [this PR comment](https://github.com/raft-tech/TANF-app/pull/941#discussion_r644045916) there is an issue with the OWASP HTML report creation which causes it to not ignore alerts that are set to be ignored in the configuration file. This ultimately causes false positives to still be present in the HTML report we capture in build artifacts and supply to OCIO periodically.

If we move to a scheduled scan against a deployed site most of these false positives will be organically not present (for example, HTTP Only site - cloud.gov always enforces HTTPS so this would not trigger). We can do this within Circle CI using [a scheduled workflow](https://circleci.com/docs/2.0/workflows/#scheduling-a-workflow) which calls a job to run the OWASP scan against a deployed site. Initially, we could target the staging site (after https://github.com/raft-tech/TANF-app/pull/977 is merged) and after we go live in production we could update it to target that site instead.

**Acceptance Criteria:** 
_Create a list of functional outcomes that must be achieved to complete this issue_

- [ ] OWASP scans get run against our Cloud.gov deployed applications on a scheduled basis
- [ ] False positives are not present in the HTML report generated by the scheduled scan
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 

- [ ] Implement a job in the Circle CI `config.yml` that does an OWASP scan against the deployed frontend and backend applications
- [ ] Create a workflow that calls the above job with a `schedule` trigger
- [ ] Document any false positives that are no longer present for scans against deployed sites
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**
- [Nightly Schedule Example](https://circleci.com/docs/2.0/workflows/#nightly-example)

**Supporting Documentation:**

**Open Questions:**

### PR DESCRIPTION
**Pull request by riatzukiza**
### Summary of Changes
Fixes #1032 
Add a job to circle CI which is triggered on at midnight UTC

### How to Test

1. Change the filter in the circle ci config to point at this branch `devops/1032/scheduled-owasp`
2. Switch the time in the config to be with in the next 30 minutes or so (the time is not exact, you will have to set it far enough in the future from your commit so the initial scans all pass before the next scheduled time)
3. Follow the links to the job

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [ ] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [ ] Check against the criteria:

As facilitator/product manager, @kniz-raft will decide if ACs are met from Raft's perspective.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)
  + [ ] Backend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Does this PR change any linting or CI settings?

### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

+ [ ] Did automated and manual testing with @iamjolly and @ttran-hub using Accessibility Insights reveal any errors introduced in this PR?
    + [See the full Accessibility Assessment plan here.](https://github.com/HHS/TANF-app/blob/main/docs/a11y/how-18f-will-test-a11y.md)

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

> NOTE: until we have a proper staging environment this may not be satisfiable prior to merging

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

+ [ ] If this PR introduces backend code, is that code documented both inline and overall?
+ [ ] If this PR introduces frontend code, is that code documented both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any security issues?

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -430,6 +441,16 @@ jobs:
           backend-appname: tdp-backend-sandbox
           frontend-appname: tdp-frontend-sandbox
 workflows:
+  nightly:
+    jobs:
+      - prod-owasp-scan
+    triggers:
+      - schedule:
+          cron: "0 0 * * *"
```
Comment (riatzukiza) [inline .circleci/config.yml:449]: set the scan to happen at midnight UTC every day

File: .circleci/config.yml [hunk 2]
```diff
@@ -430,6 +441,16 @@ jobs:
...
+          filters:
+            branches:
+              only:
+                - raft-tdp-main
```
Comment (riatzukiza) [inline .circleci/config.yml:453]: only use raft-tdp-main to run the scans from

File: tdrs-frontend/zap-scanner.sh [hunk 3]
```diff
@@ -27,8 +27,15 @@ chmod 777 $(pwd)/reports
 
 # check if running in circle CI
 
-
-if [ "$ENVIRONMENT" = "circle" ]; then
+if [ "$ENVIRONMENT" = "nightly" ]; then
```
Comment (riatzukiza) [inline tdrs-frontend/zap-scanner.sh:30]: Use the url for staging frontend to run the scan.

**Review by jtwillis92 — APPROVED** [review]
LGTM, observed output from nightly run in Circle CI and everything looks good.

**Review by abottoms-coder — APPROVED** [review]
LGTM, please schedule a tabletop at your convenience if you haven't already.

**Review by jorgegonzalez — APPROVED** [review]
LGTM 👍

File: .circleci/config.yml [hunk 4]
```diff
@@ -435,6 +446,16 @@ jobs:
           backend-appname: tdp-backend-sandbox
           frontend-appname: tdp-frontend-sandbox
 workflows:
+  nightly:
+    jobs:
+      - prod-owasp-scan
+    triggers:
+      - schedule:
+          cron: "0 0 * * *"
```
Comment (ADPennington) [inline .circleci/config.yml:454]:
```suggestion
          cron: "*/30 * * * *"
```

File: .circleci/config.yml [hunk 5]
```diff
@@ -435,6 +446,16 @@ jobs:
...
+          filters:
+            branches:
+              only:
+                - raft-tdp-main
```
Comment (ADPennington) [inline .circleci/config.yml:458]:
```suggestion
                - devops/1032/scheduled-owasp
```

File: .circleci/config.yml [hunk 6]
```diff
@@ -435,6 +452,16 @@ jobs:
           backend-appname: tdp-backend-sandbox
           frontend-appname: tdp-frontend-sandbox
 workflows:
+  nightly:
+    jobs:
+      - prod-owasp-scan
+    triggers:
+      - schedule:
+          cron: "50 16 * * *"
```
Comment (ADPennington) [inline .circleci/config.yml:460]:
```suggestion
          cron: "47 18* * *"
```

File: .circleci/config.yml [hunk 7]
```diff
@@ -435,6 +452,16 @@ jobs:
           backend-appname: tdp-backend-sandbox
           frontend-appname: tdp-frontend-sandbox
 workflows:
+  nightly:
+    jobs:
+      - prod-owasp-scan
+    triggers:
+      - schedule:
+          cron: "47 18* * *"
```
Comment (ADPennington) [inline .circleci/config.yml:460]:
```suggestion
          cron: "47 18 * * *"
```

**Comment by ADPennington** [comment]
@riatzukiza frontend report generated (https://8904-281707402-gh.circle-artifacts.com/0/tdrs-frontend/reports/owasp_report.html) but not the backend (https://app.circleci.com/pipelines/github/raft-tech/TANF-app/4276/workflows/d6e91285-c3b7-46e6-9477-7f1a71c4da92/jobs/8904/parallel-runs/0/steps/0-105)

**Comment by riatzukiza** [comment]
@ADPennington All the issues should be dealt with, I refactored the zap scanner logic to use one script, several of these issues seemed to me to have come up from code duplication, and me not dealing with both files each time I made a change.

File: scripts/zap-scanner.sh [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+echo $REPORTS_DIR
+
+echo $APP_URL
+echo $CONFIG_FILE
+
+if command -v sha256sum >/dev/null ; then
+	SHASUM=sha256sum
+elif command -v shasum >/dev/null ; then
+	SHASUM=shasum
+fi
```
Comment (jtwillis92) [inline scripts/zap-scanner.sh:57]: This block appears to be unused.

Comment (riatzukiza) [reply scripts/zap-scanner.sh:57]: Thats what I thought as well.

**Comment by ADPennington** [comment]
test results: 👍🏾 

- Frontend vulnerabilities removed from html rpt w/ scan against deployed site:
  - Medium:
    - HTTP only site (10106)
    - X-frame-options header not set (10020)
  - Low:
    - Server leaks (10036)
    - X-content-type options header missing (10021)
   
- Backend:
   - Medium:
     - HTTP only site (10106)
  - Low:
    - Server leaks (10036)

File: .circleci/config.yml [hunk 9]
```diff
@@ -435,6 +452,16 @@ jobs:
           backend-appname: tdp-backend-sandbox
           frontend-appname: tdp-frontend-sandbox
 workflows:
+  nightly:
+    jobs:
+      - prod-owasp-scan
+    triggers:
+      - schedule:
+          cron: "47 18 * * *"
```
Comment (ADPennington) [inline .circleci/config.yml:460]:
```suggestion
          cron: "* * * * *"
```

File: .circleci/config.yml [hunk 10]
```diff
@@ -435,6 +452,16 @@ jobs:
...
+          filters:
+            branches:
+              only:
+                - devops/1032/scheduled-owasp
```
Comment (ADPennington) [inline .circleci/config.yml:464]:
```suggestion
                - raft-tdp-main
```

File: scripts/zap-scanner.sh (hunk 8 above)
Comment (ADPennington) [reply scripts/zap-scanner.sh:57]: @riatzukiza @jtwillis92 how do you want to handle this? create a ticket to clean it up? add it as a task to another related ticket or fix here?

**Review by ADPennington — APPROVED** [review]
@abottoms-coder this is ready to merge. thanks @riatzukiza 👍🏾 

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [x] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [x] Check against the criteria:
  + [x] OWASP scans get run against our Cloud.gov deployed applications on a scheduled basis
  + [x] False positives are not present in the HTML report generated by the scheduled scan
https://github.com/raft-tech/TANF-app/pull/1103#issuecomment-888564176
 + [x] Testing Checklist has been run and all tests pass
As Product Owner, @lfrohlich will decide if ACs are met.

### Deliverable 2: Tested Code

_n/a: manually tested the scheduled scan_ 

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Does this PR change any linting or CI settings?

### Deliverable 4: Accessible

_n/a_

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

_n/a_

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any security issues?
one new [issue](https://9055-281707402-gh.circle-artifacts.com/0/tdrs-frontend/reports/owasp_report.html#10015) that can be addressed as part of a followup ticket, if needed. we werent able to see this before since we werent scanning against deployed site. cc: @abottoms-coder

File: scripts/zap-scanner.sh (hunk 8 above)
Comment (jtwillis92) [reply scripts/zap-scanner.sh]: Looks like it was removed in a recent commit, which I agree is the correct move.

**Pull request merged**